src/main/java/org/myrobotlab/codec/Api.java

```diff
@@ -142,21 +158,214 @@ static public Message uriToMsg(String uri) {
+    r.setBroadcaster(webgui.getBroadcaster());
+  }
+
+ 
+  protected void setHeaderContentType(WebGui webgui, String apiKey, AtmosphereResource r) {
+    r.getResponse().addHeader("Content-Type", CodecUtils.MIME_TYPE_JSON);
+  }
+
+  protected void suspend(WebGui webgui, String apiKey, AtmosphereResource r) {
+    if (!apiKey.equals(ApiFactory.API_TYPE_SERVICE) && !r.isSuspended()) {
+      r.suspend();
+    }
+  }
+
+  protected void handleError(WebGui webgui, String apiKey, AtmosphereResource r, Exception e) {
+    
+    try {
+
+      AtmosphereResponse response = r.getResponse();
+      OutputStream out = response.getOutputStream();
+
+      response.addHeader("Content-Type", CodecUtils.MIME_TYPE_JSON);
+      Codec codec = CodecFactory.getCodec(CodecUtils.MIME_TYPE_JSON);
+
+      Status error = Status.error(e);   // name == null anonymous ?
+      Message msg = Message.createMessage(webgui, null, CodecUtils.getCallbackTopicName("getStatus"), error);
+      if (ApiFactory.API_TYPE_SERVICE.equals(apiKey)) {
+        // for the purpose of only returning the data
+        // e.g. http://api/services/runtime/getUptime -> return the uptime
+        // only not the message
+        if (msg.data == null) {
+          log.warn("<< {}", CodecJson.encode(null));
+          codec.encode(out, null);
+        } else {
+          // return the return type
+          log.warn("<< {}", CodecJson.encode(msg.data[0]));
+          codec.encode(out, msg.data[0]);
+        }
+      } else {
+        // API_TYPE_MESSAGES
+        // DEPRECATE - FOR LOGGING ONLY REMOVE
+        log.warn("<< {}", CodecJson.encode(msg)); // FIXME if logTraffic
+        codec.encode(out, msg);
+      }
+
+    } catch (Exception e2) {
+      log.error("respond threw", e);
+    }
+
+  }
+
+  // FIXME - max complexity !!!!
+  // abstract public Object process(WebGui webgui, OutputStream out, Message msgFromUri, String data) throws Exception;
+
+  /**
+   * High level minimal parameter process for AtmosphereResource
+   * Many of the details of the flow control (suspend, broadcast, error handling) are dependent on the
+   * api being used
+   * @param webgui
+   * @param apiKey
+   * @param r
+   */
+  public Object process(WebGui webgui, String apiKey, AtmosphereResource r) {
+    
+    try {
+    
+    setBroadcaster(webgui, apiKey, r);
+    handleSession(webgui, apiKey, r);
+    suspend(webgui, apiKey, r);
+    setHeaderContentType(webgui, apiKey, r);
+    String data = r.getRequest().body().asString();
+ 
+    // CALLING MAX COMPLEXITY
+    return process(webgui, apiKey, r.getRequest().getRequestURI(), r.getResponse().getOutputStream(), data);
+    
+    } catch(Exception e) {
+      handleError(webgui, apiKey, r, e);
+      return null;
+    }
+  }
+  
+  // MAX COMPLEXITY WITH SIMPLEST DATA TYPES
+  protected Object process(MessageSender sender, String apiKey, String uri, OutputStream out, String json) throws Exception {

```

Comment:
This method doesn't do anything, only returns null..  is it needed?

Reply:
removed locally - will be in next pr